### DISCUSSION TITLE
How would you check equality between a compile-time protocol reference and a conformance?

### DISCUSSION THREAD

**Discussion by NSExceptional** [Azoy/Echo#35] [Q&A] [answered]
i.e. you want to check if `Equatable.self` is present in `metadata.conformances`. Can't seem to wrap my head around how to compare the two besides comparing the names which is not intuitively reliable

**Comment by Azoy** [comment, marked as answer]
So doing something like `reflect(Equatable.self)` isn't possible yet (soon it will be because of https://forums.swift.org/t/se-0309-unlock-existential-types-for-all-protocols/47515 !), but if one really wants to do this kind of operation now you could do something pretty hacky.

```swift
let equatableMetadata = reflect(_typeByName("SQ")!) as! ExistentialMetadata
let equatable = equatableMetadata.protocols[0]

for conformance in metadata.conformances {
  if conformance.protocol == equatable { /* ... */ }
}
```

I'm thinking I may want to make the conformances a dictionary with the protocol as the key instead. Keep in mind that `ExistentialMetadata` is not ABI stable so it can change in the future, but for experimenting purposes this works right now.

**Reply by NSExceptional** [reply]
Ugh, of course that was conveniently left out when we got ABI stability ;(

Thanks!

Edit: for anyone reading this, to make it work with `SomeProtocol.self`, you need to do this:

```swift
extension TypeMetadata {
    func conforms(to _protocol: Any) -> Bool {
        let metatype = reflect(_protocol) as! MetatypeMetadata
        let existential = metatype.instanceMetadata as! ExistentialMetadata
        let desc = existential.protocols.first!

        return !conformances.filter({ $0.protocol == desc }).isEmpty
    }
}

let conforms = someTypeMetadata.conforms(to: SomeProtocol.self)
```

**Reply by NSExceptional** [reply]
Do you think this sort of check is something you would want to provide in Echo itself? Or at a higher level in another library instead?
